### ISSUE TITLE
Add new bombs to Bombsquad

### ISSUE THREAD

**Issue by Gengar-lab** [Mrmaxmeier/BombSquad-Community-Mod-Manager#69]
Like snowypowerup.py, can anyone tell how to add new bombs to bombsquad without changing internal scripts as snowypowerup added powerups.

Please also make a new bomb and show.

**Comment by rahulraman0108** [comment]
Well, @Benefit-Zebra thank you for this idea, I think this idea is very good, and I will start to work on this idea when I get time because currently, I am a little bit busy in my school exams. I promise that I will notify you ASAP after I come to the result of my work on this idea.

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, what type of bomb should I make?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great My friend suggested this idea of what bomb to make:

Please make a bomb which when thrown, goes up slowly and search for player of opposite team or any other player in FFA. And when bomb finds a player , it will stay in  touch with player(not stick to the player ; like the bomb will surround his body).

The bomb will have much longer fuse time and can get blast on punching or blasting bomb near it (The bomb will not get blasted on touching bomb of his own type, i.e, If a player throws three bomb , it will surround another player and the bomb will not get blast on touching each other.It will only get blast by throwing another bomb eg. impact bomb on it).

**Comment by rahulraman0108** [comment]
Okay @Benefit-Zebra, I will start the work tonight at 7:30 PM according to GMT+5:30 and try to finish it ASAP, first I will make a new normal impact bomb which on exploding gives out 1-3 random powerups to see if it works then I will start to make the bomb as you have stated, and due to my exams you may need to wait for 2 to 4 days, so please be patient.

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

I know of exams as I also stay in same country.I will be patient. Please inform me when you finish it.

Thanks for making bomb which my friend suggested.

**Comment by rahulraman0108** [comment]
Oh, I did not know that, bye, exam's time now.
And welcome.

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, I made a test bomb as I stated just now, and it worked, now I will start making a bomb from your idea, it may take some time so please try to be patient.
Do you want some photos and the code file?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

I will be patient as your exams are going on.

Yess I would like to see photos and want code file to test it.

Please inform me when making of bomb will be finished.

I also wanted to know that where is "Meteoroid Rampage Special" Map is located in OSMP Modpack code file scripts.

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, I will include the link of the video and files of the test impact bomb I made in this comment when it gets uploaded on MediaFire, I will inform you ASAP after the work of making the bomb of your idea is completed, and the code for that map in OSMP is in MsV6.py file.

And thank you because your question reminded me to make OSMP1.2, I just forgot of it.

And a short note of OSMP:
    OSMP1.0 was just a test modpack made by me it hass less original content and more content from github or other programmers, well in OSMP1.2 mostly there will be my mods and my adds and some permitted mods with copyright mention, there are many issues with OSMP1.0 which I need to fix in 1.2.

_Currently going school for exam starting in an hour, I will upload the video at night when father comes back from office then update this comment.
Have a good day._

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

All the best for your upcoming exams.

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, I have completed more than half of the work of the video, now I need to test the code and then only I can proceed, but I can not test that in one player and I do not have a controller or any other device to play BombSquad, what do I do, any idea?

**Comment by rahulraman0108** [comment]
@Benefit-Zebra I have an idea, I host a server and in I that put the files, then when you get ready I send you the IP and port and we both test. Comment here or mail me when you get ready.

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

Today I am busy with function whole day.

Tomorrow I can only be online at night when my father comes home (around 9:45 pm.)

I have idea , first go in Setting > Advanced >Benchmark and Stress-Test and then click on Run stress test.Automatically some players as test inputs will get connected.And then try bomb on them.
OR
Try the bombs on bots in custom mods in which enemies come.

**Comment by rahulraman0108** [comment]
Well, okay I will do that, thanks for this.

**Comment by Benefit-Zebra** [comment]
@I-Am-The-Great

Relative's Marriage

Please inform so I delete this comment as you read it.

**Comment by I-Am-The-Great** [comment]
@Benefit-Zebra I have read

**Comment by rahulraman0108** [comment]
@Benefit-Zebra I made the bomb, now making its video then I will post the video on youtube and share the link here.
Features of the bomb are:
1. Automatically aims to an opponent (only opponent no teammate)
2. Blasts on being punched
3. Blasts if touches another bomb of another kind, not its same bomb type.
4. Longer fuse time: 13 seconds
5. Roam over the head of the target, and not stick to it and fly
The bomb's file's code:
```python
import bs
import random
import bsBomb
import bsUtils
from bsBomb import Bomb, ExplodeMessage, ArmMessage, WarnMessage, Blast, BombFactory, \
    ExplodeHitMessage, ImpactMessage, SplatMessage


class AimForOpponent(object):
    def __init__(self, bomb, owner):

        self.bomb = bomb
        self.owner = owner
        self.target = None

        self.aimZoneSpaz = bs.Material()
        self.aimZoneSpaz.addActions(conditions=(('theyHaveMaterial', bs.getSharedObject('playerMaterial'))),
                                    actions=(("modifyPartCollision", "collide", True),
                                             ("modifyPartCollision", "physical", False),
                                             ("call", "atConnect", self.touchedSpaz)))

        self.lookForSpaz()

    def lookForSpaz(self):
        # To slow down the movement of the bomb towards the target and put the bomb over the head of the target.
        self.bomb.extraAcceleration = (0, 20, 0)
        self.node = bs.newNode('region',
                               attrs={'position': (self.bomb.position[0], self.bomb.position[1],
                                                   self.bomb.position[2]) if self.bomb.exists() else (
                                   0, 0, 0),
                                      'scale': (0.0, 0.0, 0.0),
                                      'type': 'sphere',
                                      'materials': [self.aimZoneSpaz]})
        self.s = bsUtils.animateArray(self.node, "scale", 3, {0: (0.0, 0.0, 0.0), 50: (60, 60, 60), 100: (90, 90, 90)},
                                      True)

        bs.gameTimer(150, self.node.delete)

        def checkTarget():
            if self.target is not None:
                self.touchedSpaz()

        bs.gameTimer(151, checkTarget)

    def go(self):
        if self.target is not None and self.bomb is not None and self.bomb.exists():
            self.bomb.velocity = (
                self.bomb.velocity[0] + (self.target.position[0] - self.bomb.position[0]),
                self.bomb.velocity[1] + (self.target.position[1] - self.bomb.position[1]),
                self.bomb.velocity[2] + (self.target.position[2] - self.bomb.position[2]))
            bs.gameTimer(1, self.go)

    def touchedSpaz(self):
        node = bs.getCollisionInfo('opposingNode')
        if not node == self.owner and node.getDelegate().isAlive() \
                and node.getDelegate().getPlayer().getTeam() != self.owner.getDelegate().getPlayer().getTeam():
            self.target = node
            self.s = None
            self.node.delete()
            self.bomb.extraAcceleration = (0, 200, 0)
            self.go()


class NewBombFactory(BombFactory):
    def __init__(self):
        self.bombModel = bs.getModel('bomb')
        self.stickyBombModel = bs.getModel('bombSticky')
        self.impactBombModel = bs.getModel('impactBomb')
        self.landMineModel = bs.getModel('landMine')
        self.tntModel = bs.getModel('tnt')

        self.regularTex = bs.getTexture('bombColor')
        self.iceTex = bs.getTexture('bombColorIce')
        self.stickyTex = bs.getTexture('bombStickyColor')
        self.impactTex = bs.getTexture('impactBombColor')
        self.impactLitTex = bs.getTexture('impactBombColorLit')
        self.landMineTex = bs.getTexture('landMine')
        self.landMineLitTex = bs.getTexture('landMineLit')
        self.tntTex = bs.getTexture('tnt')
        self.newTex = bs.getTexture("achievementOnslaught")

        self.hissSound = bs.getSound('hiss')
        self.debrisFallSound = bs.getSound('debrisFall')
        self.woodDebrisFallSound = bs.getSound('woodDebrisFall')

        self.explodeSounds = (bs.getSound('explosion01'),
                              bs.getSound('explosion02'),
                              bs.getSound('explosion03'),
                              bs.getSound('explosion04'),
                              bs.getSound('explosion05'))

        self.freezeSound = bs.getSound('freeze')
        self.fuseSound = bs.getSound('fuse01')
        self.activateSound = bs.getSound('activateBeep')
        self.warnSound = bs.getSound('warnBeep')

        # set up our material so new bombs dont collide with objects
        # that they are initially overlapping
        self.bombMaterial = bs.Material()
        self.normalSoundMaterial = bs.Material()
        self.stickyMaterial = bs.Material()

        self.bombMaterial.addActions(
            conditions=((('weAreYoungerThan', 100),
                         'or', ('theyAreYoungerThan', 100)),
                        'and', ('theyHaveMaterial',
                                bs.getSharedObject('objectMaterial'))),
            actions=(('modifyNodeCollision', 'collide', False)))

        # we want pickup materials to always hit us even if we're currently not
        # colliding with their node (generally due to the above rule)
        self.bombMaterial.addActions(
            conditions=('theyHaveMaterial',
                        bs.getSharedObject('pickupMaterial')),
            actions=(('modifyPartCollision', 'useNodeCollide', False)))

        self.bombMaterial.addActions(actions=('modifyPartCollision',
                                              'friction', 0.3))

        self.landMineNoExplodeMaterial = bs.Material()
        self.landMineBlastMaterial = bs.Material()
        self.landMineBlastMaterial.addActions(
            conditions=(
                ('weAreOlderThan', 200),
                'and', ('theyAreOlderThan', 200),
                'and', ('evalColliding',),
                'and', (('theyDontHaveMaterial',
                         self.landMineNoExplodeMaterial),
                        'and', (('theyHaveMaterial',
                                 bs.getSharedObject('objectMaterial')),
                                'or', ('theyHaveMaterial',
                                       bs.getSharedObject('playerMaterial'))))),
            actions=(('message', 'ourNode', 'atConnect', ImpactMessage())))

        self.impactBlastMaterial = bs.Material()
        self.impactBlastMaterial.addActions(
            conditions=(('weAreOlderThan', 200),
                        'and', ('theyAreOlderThan', 200),
                        'and', ('evalColliding',),
                        'and', (('theyHaveMaterial',
                                 bs.getSharedObject('footingMaterial')),
                                'or', ('theyHaveMaterial',
                                       bs.getSharedObject('objectMaterial')))),
            actions=(('message', 'ourNode', 'atConnect', ImpactMessage())))

        self.newImpactBlastMaterial = bs.Material()
        self.newImpactBlastMaterial.addActions(
            conditions=(('weAreOlderThan', 200),
                        'and', ('theyAreOlderThan', 200),
                        'and', ('evalColliding',),
                        'and', ('theyDontHaveMaterial', bs.getSharedObject('playerMaterial')),
                        'and', ('theyHaveMaterial', self.bombMaterial),
                        'and', ('theyDontHaveMaterial', self.newImpactBlastMaterial)),
            actions=(('message', 'ourNode', 'atConnect', ImpactMessage())))

        self.blastMaterial = bs.Material()
        self.blastMaterial.addActions(
            conditions=(('theyHaveMaterial',
                         bs.getSharedObject('objectMaterial'))),
            actions=(('modifyPartCollision', 'collide', True),
                     ('modifyPartCollision', 'physical', False),
                     ('message', 'ourNode', 'atConnect', ExplodeHitMessage())))

        self.dinkSounds = (bs.getSound('bombDrop01'),
                           bs.getSound('bombDrop02'))
        self.stickyImpactSound = bs.getSound('stickyImpact')
        self.rollSound = bs.getSound('bombRoll01')

        # collision sounds
        self.normalSoundMaterial.addActions(
            conditions=('theyHaveMaterial',
                        bs.getSharedObject('footingMaterial')),
            actions=(('impactSound', self.dinkSounds, 2, 0.8),
                     ('rollSound', self.rollSound, 3, 6)))

        self.stickyMaterial.addActions(
            actions=(('modifyPartCollision', 'stiffness', 0.1),
                     ('modifyPartCollision', 'damping', 1.0)))

        self.stickyMaterial.addActions(
            conditions=(('theyHaveMaterial',
                         bs.getSharedObject('playerMaterial')),
                        'or', ('theyHaveMaterial',
                               bs.getSharedObject('footingMaterial'))),
            actions=(('message', 'ourNode', 'atConnect', SplatMessage())))


class NewBlast(Blast):
    def __init__(self, position=(0, 1, 0), velocity=(0, 0, 0), blastRadius=2.0,
                 blastType="normal", sourcePlayer=None, hitType='explosion',
                 hitSubType='normal'):
        bs.Actor.__init__(self)

        factory = Bomb.getFactory()

        self.blastType = blastType
        self.sourcePlayer = sourcePlayer

        self.hitType = hitType;
        self.hitSubType = hitSubType;

        # blast radius
        self.radius = blastRadius

        # set our position a bit lower so we throw more things upward
        self.node = bs.newNode('region', delegate=self, attrs={
            'position': (position[0], position[1] - 0.1, position[2]),
            'scale': (self.radius, self.radius, self.radius),
            'type': 'sphere',
            'materials': (factory.blastMaterial,
                          bs.getSharedObject('attackMaterial'))})

        bs.gameTimer(50, self.node.delete)

        # throw in an explosion and flash
        explosion = bs.newNode("explosion", attrs={
            'position': position,
            'velocity': (velocity[0], max(-1.0, velocity[1]), velocity[2]),
            'radius': self.radius,
            'big': (self.blastType == 'tnt')})
        if self.blastType == "ice":
            explosion.color = (0, 0.05, 0.4)

        bs.gameTimer(1000, explosion.delete)

        if self.blastType != 'ice':
            bs.emitBGDynamics(position=position, velocity=velocity,
                              count=int(1.0 + random.random() * 4),
                              emitType='tendrils', tendrilType='thinSmoke')
        bs.emitBGDynamics(
            position=position, velocity=velocity,
            count=int(4.0 + random.random() * 4), emitType='tendrils',
            tendrilType='ice' if self.blastType == 'ice' else 'smoke')
        bs.emitBGDynamics(
            position=position, emitType='distortion',
            spread=1.0 if self.blastType == 'tnt' else 2.0)

        # and emit some shrapnel..
        if self.blastType == 'ice':
            def _doEmit():
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=30, spread=2.0, scale=0.4,
                                  chunkType='ice', emitType='stickers');

            bs.gameTimer(50, _doEmit)  # looks better if we delay a bit

        elif self.blastType == 'sticky':
            def _doEmit():
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(4.0 + random.random() * 8),
                                  spread=0.7, chunkType='slime');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(4.0 + random.random() * 8), scale=0.5,
                                  spread=0.7, chunkType='slime');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=15, scale=0.6, chunkType='slime',
                                  emitType='stickers');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=20, scale=0.7, chunkType='spark',
                                  emitType='stickers');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(6.0 + random.random() * 12),
                                  scale=0.8, spread=1.5, chunkType='spark');

            bs.gameTimer(50, _doEmit)  # looks better if we delay a bit

        elif self.blastType == 'impact':  # regular bomb shrapnel
            def _doEmit():
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(4.0 + random.random() * 8), scale=0.8,
                                  chunkType='metal');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(4.0 + random.random() * 8), scale=0.4,
                                  chunkType='metal');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=20, scale=0.7, chunkType='spark',
                                  emitType='stickers');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(8.0 + random.random() * 15), scale=0.8,
                                  spread=1.5, chunkType='spark');

            bs.gameTimer(50, _doEmit)  # looks better if we delay a bit

        elif self.blastType == 'new':  # regular bomb shrapnel

            def _doEmit():
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(4.0 + random.random() * 8), scale=0.8,
                                  chunkType='metal');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(4.0 + random.random() * 8), scale=0.4,
                                  chunkType='metal');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=20, scale=0.7, chunkType='spark',
                                  emitType='stickers');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(8.0 + random.random() * 15), scale=0.8,
                                  spread=1.5, chunkType='spark');

            bs.gameTimer(50, _doEmit)  # looks better if we delay a bit

        else:  # regular or land mine bomb shrapnel
            def _doEmit():
                if self.blastType != 'tnt':
                    bs.emitBGDynamics(position=position, velocity=velocity,
                                      count=int(4.0 + random.random() * 8),
                                      chunkType='rock');
                    bs.emitBGDynamics(position=position, velocity=velocity,
                                      count=int(4.0 + random.random() * 8),
                                      scale=0.5, chunkType='rock');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=30,
                                  scale=1.0 if self.blastType == 'tnt' else 0.7,
                                  chunkType='spark', emitType='stickers');
                bs.emitBGDynamics(position=position, velocity=velocity,
                                  count=int(18.0 + random.random() * 20),
                                  scale=1.0 if self.blastType == 'tnt' else 0.8,
                                  spread=1.5, chunkType='spark');

                # tnt throws splintery chunks
                if self.blastType == 'tnt':
                    def _emitSplinters():
                        bs.emitBGDynamics(position=position, velocity=velocity,
                                          count=int(20.0 + random.random() * 25),
                                          scale=0.8, spread=1.0,
                                          chunkType='splinter');

                    bs.gameTimer(10, _emitSplinters)

                # every now and then do a sparky one
                if self.blastType == 'tnt' or random.random() < 0.1:
                    def _emitExtraSparks():
                        bs.emitBGDynamics(position=position, velocity=velocity,
                                          count=int(10.0 + random.random() * 20),
                                          scale=0.8, spread=1.5,
                                          chunkType='spark');

                    bs.gameTimer(20, _emitExtraSparks)

            bs.gameTimer(50, _doEmit)  # looks better if we delay a bit

        light = bs.newNode('light', attrs={
            'position': position,
            'volumeIntensityScale': 10.0,
            'color': ((0.6, 0.6, 1.0) if self.blastType == 'ice'
                      else (1, 0.3, 0.1))})

        s = random.uniform(0.6, 0.9)
        scorchRadius = lightRadius = self.radius
        if self.blastType == 'tnt':
            lightRadius *= 1.4
            scorchRadius *= 1.15
            s *= 3.0

        iScale = 1.6
        bsUtils.animate(light, "intensity", {
            0: 2.0 * iScale, int(s * 20): 0.1 * iScale,
            int(s * 25): 0.2 * iScale, int(s * 50): 17.0 * iScale, int(s * 60): 5.0 * iScale,
            int(s * 80): 4.0 * iScale, int(s * 200): 0.6 * iScale,
            int(s * 2000): 0.00 * iScale, int(s * 3000): 0.0})
        bsUtils.animate(light, "radius", {
            0: lightRadius * 0.2, int(s * 50): lightRadius * 0.55,
            int(s * 100): lightRadius * 0.3, int(s * 300): lightRadius * 0.15,
            int(s * 1000): lightRadius * 0.05})
        bs.gameTimer(int(s * 3000), light.delete)

        # make a scorch that fades over time
        scorch = bs.newNode('scorch', attrs={
            'position': position,
            'size': scorchRadius * 0.5,
            'big': (self.blastType == 'tnt')})
        if self.blastType == 'ice':
            scorch.color = (1, 1, 1.5)

        bsUtils.animate(scorch, "presence", {3000: 1, 13000: 0})
        bs.gameTimer(13000, scorch.delete)

        if self.blastType == 'ice':
            bs.playSound(factory.hissSound, position=light.position)

        p = light.position
        bs.playSound(factory.getRandomExplodeSound(), position=p)
        bs.playSound(factory.debrisFallSound, position=p)

        bs.shakeCamera(intensity=5.0 if self.blastType == 'tnt' else 1.0)

        # tnt is more epic..
        if self.blastType == 'tnt':
            bs.playSound(factory.getRandomExplodeSound(), position=p)

            def _extraBoom():
                bs.playSound(factory.getRandomExplodeSound(), position=p)

            bs.gameTimer(250, _extraBoom)

            def _extraDebrisSound():
                bs.playSound(factory.debrisFallSound, position=p)
                bs.playSound(factory.woodDebrisFallSound, position=p)

            bs.gameTimer(400, _extraDebrisSound)


class NewBomb(Bomb):
    def __init__(self, position=(0, 1, 0), velocity=(0, 0, 0), bombType='normal',
                 blastRadius=2.0, sourcePlayer=None, owner=None):
        bs.Actor.__init__(self)

        factory = self.getFactory()

        if not bombType in ('new', 'ice', 'impact', 'landMine', 'normal', 'sticky', 'tnt'):
            raise Exception("invalid bomb type: " + bombType)
        self.bombType = bombType

        self._exploded = False

        if self.bombType == 'sticky': self._lastStickySoundTime = 0

        self.blastRadius = blastRadius
        if self.bombType == 'ice':
            self.blastRadius *= 1.2
        elif self.bombType == 'impact':
            self.blastRadius *= 0.7
        elif self.bombType == 'landMine':
            self.blastRadius *= 0.7
        elif self.bombType == 'tnt':
            self.blastRadius *= 1.45

        self._explodeCallbacks = []

        # the player this came from
        self.sourcePlayer = sourcePlayer

        # by default our hit type/subtype is our own, but we pick up types of
        # whoever sets us off so we know what caused a chain reaction
        self.hitType = 'explosion'
        self.hitSubType = self.bombType

        # if no owner was provided, use an unconnected node ref
        if owner is None: owner = bs.Node(None)

        # the node this came from
        self.owner = owner

        # adding footing-materials to things can screw up jumping and flying
        # since players carrying those things
        # and thus touching footing objects will think they're on solid ground..
        # perhaps we don't wanna add this even in the tnt case?..
        if self.bombType == 'tnt':
            materials = (factory.bombMaterial,
                         bs.getSharedObject('footingMaterial'),
                         bs.getSharedObject('objectMaterial'))
        else:
            materials = (factory.bombMaterial,
                         bs.getSharedObject('objectMaterial'))

        if self.bombType == 'impact':
            materials = materials + (factory.impactBlastMaterial,)
        elif self.bombType == 'new':
            materials = materials + (factory.newImpactBlastMaterial,)
        elif self.bombType == 'landMine':
            materials = materials + (factory.landMineNoExplodeMaterial,)

        if self.bombType == 'sticky':
            materials = materials + (factory.stickyMaterial,)
        else:
            materials = materials + (factory.normalSoundMaterial,)

        if self.bombType == 'landMine':
            self.node = bs.newNode('prop', delegate=self, attrs={
                'position': position,
                'velocity': velocity,
                'model': factory.landMineModel,
                'lightModel': factory.landMineModel,
                'body': 'landMine',
                'shadowSize': 0.44,
                'colorTexture': factory.landMineTex,
                'reflection': 'powerup',
                'reflectionScale': [1.0],
                'materials': materials})

        elif self.bombType == 'tnt':
            self.node = bs.newNode('prop', delegate=self, attrs={
                'position': position,
                'velocity': velocity,
                'model': factory.tntModel,
                'lightModel': factory.tntModel,
                'body': 'crate',
                'shadowSize': 0.5,
                'colorTexture': factory.tntTex,
                'reflection': 'soft',
                'reflectionScale': [0.23],
                'materials': materials})

        elif self.bombType == 'impact':
            fuseTime = 20000
            self.node = bs.newNode('prop', delegate=self, attrs={
                'position': position,
                'velocity': velocity,
                'body': 'sphere',
                'model': factory.impactBombModel,
                'shadowSize': 0.3,
                'colorTexture': factory.impactTex,
                'reflection': 'powerup',
                'reflectionScale': [1.5],
                'materials': materials})
            self.armTimer = bs.Timer(200, bs.WeakCall(self.handleMessage,
                                                      ArmMessage()))
            self.warnTimer = bs.Timer(fuseTime - 1700,
                                      bs.WeakCall(self.handleMessage,
                                                  WarnMessage()))

        else:
            fuseTime = 3000
            if self.bombType == 'sticky':
                sticky = True
                model = factory.stickyBombModel
                rType = 'sharper'
                rScale = 1.8
            else:
                sticky = False
                model = factory.bombModel
                rType = 'sharper'
                rScale = 1.8
            if self.bombType == 'ice':
                tex = factory.iceTex
            elif self.bombType == 'sticky':
                tex = factory.stickyTex
            elif self.bombType == 'new':
                fuseTime = 13000
                tex = factory.newTex
                model = factory.impactBombModel
            else:
                tex = factory.regularTex
            self.node = bs.newNode('bomb', delegate=self, attrs={
                'position': position,
                'velocity': velocity,
                'model': model,
                'shadowSize': 0.3,
                'colorTexture': tex,
                'sticky': sticky,
                'owner': owner,
                'reflection': rType,
                'reflectionScale': [rScale],
                'materials': materials})

            sound = bs.newNode('sound', owner=self.node, attrs={
                'sound': factory.fuseSound,
                'volume': 0.25})
            self.node.connectAttr('position', sound, 'position')
            bsUtils.animate(self.node, 'fuseLength', {0: 1.0, fuseTime: 0.0})

        # light the fuse!!!
        if self.bombType not in ('landMine', 'tnt'):
            bs.gameTimer(fuseTime,
                         bs.WeakCall(self.handleMessage, ExplodeMessage()))

        bsUtils.animate(self.node, "modelScale", {0: 0, 200: 1.3, 260: 1})

    def _handleDropped(self, m):
        if self.bombType == 'landMine':
            self.armTimer = \
                bs.Timer(1250, bs.WeakCall(self.handleMessage, ArmMessage()))

        # once we've thrown a sticky bomb we can stick to it..
        elif self.bombType == 'sticky':
            def _safeSetAttr(node, attr, value):
                if node.exists(): setattr(node, attr, value)

            bs.gameTimer(
                250, lambda: _safeSetAttr(self.node, 'stickToOwner', True))

        elif self.bombType == 'new':
            AimForOpponent(self.node, self.owner)

    def _handleHit(self, msg):
        if self.bombType == 'new' and msg.hitType == 'punch':
            self.handleMessage(ExplodeMessage())
        isPunch = (msg.srcNode.exists() and msg.srcNode.getNodeType() == 'spaz')

        # normal bombs are triggered by non-punch impacts..
        # impact-bombs by all impacts

        if (not self._exploded and not isPunch
                or self.bombType in ['impact', 'landMine']):
            # also lets change the owner of the bomb to whoever is setting
            # us off.. (this way points for big chain reactions go to the
            # person causing them)
            if msg.sourcePlayer not in [None]:
                self.sourcePlayer = msg.sourcePlayer

                # also inherit the hit type (if a landmine sets off by a bomb,
                # the credit should go to the mine)
                # the exception is TNT.  TNT always gets credit.
                if self.bombType != 'tnt':
                    self.hitType = msg.hitType
                    self.hitSubType = msg.hitSubType

            bs.gameTimer(100 + int(random.random() * 100),
                         bs.WeakCall(self.handleMessage, ExplodeMessage()))
        self.node.handleMessage(
            "impulse", msg.pos[0], msg.pos[1], msg.pos[2],
            msg.velocity[0], msg.velocity[1], msg.velocity[2],
            msg.magnitude, msg.velocityMagnitude, msg.radius, 0,
            msg.velocity[0], msg.velocity[1], msg.velocity[2])

        if msg.srcNode.exists():
            pass

    def _handleImpact(self, m):
        node, body = bs.getCollisionInfo("opposingNode", "opposingBody")
        # if we're an impact bomb and we came from this node, don't explode...
        # alternately if we're hitting another impact-bomb from the same source,
        # don't explode...
        try:
            nodeDelegate = node.getDelegate()
        except Exception:
            nodeDelegate = None
        if node is not None and node.exists():
            if (self.bombType == 'impact' and
                    (node is self.owner
                     or (isinstance(nodeDelegate, Bomb)
                         and nodeDelegate.bombType == 'impact'
                         and nodeDelegate.owner is self.owner))):
                return
            else:
                self.handleMessage(ExplodeMessage())

    def handleMessage(self, msg):
        if isinstance(msg, ExplodeMessage):
            self.explode()
        elif isinstance(msg, ImpactMessage):
            self._handleImpact(msg)
        elif isinstance(msg, bs.PickedUpMessage):
            # change our source to whoever just picked us up *only* if its None
            # this way we can get points for killing bots with their own bombs
            # hmm would there be a downside to this?...
            if self.sourcePlayer is not None:
                self.sourcePlayer = msg.node.sourcePlayer
        elif isinstance(msg, SplatMessage):
            self._handleSplat(msg)
        elif isinstance(msg, bs.DroppedMessage):
            self._handleDropped(msg)
        elif isinstance(msg, bs.HitMessage):
            self._handleHit(msg)
        elif isinstance(msg, bs.DieMessage):
            self._handleDie(msg)
        elif isinstance(msg, bs.OutOfBoundsMessage):
            self._handleOOB(msg)
        elif isinstance(msg, ArmMessage):
            self.arm()
        elif isinstance(msg, WarnMessage):
            self._handleWarn(msg)
        else:
            bs.Actor.handleMessage(self, msg)


bsBomb.Bomb = NewBomb
bs.Bomb = NewBomb
bsBomb.Blast = NewBlast
bs.Blast = NewBlast
bsBomb.BombFactory = NewBombFactory
bs.BombFactory = NewBombFactory
```
The bomb's powerup's file's code:
```python
import bs
import random
import bsPowerup
import bsSpaz
from bsPowerup import PowerupMessage, PowerupAcceptMessage, _TouchedMessage, PowerupFactory, Powerup

defaultPowerupInterval = 8000


class NewPowerupFactory(PowerupFactory):
    def __init__(self):
        self._lastPowerupType = None

        self.model = bs.getModel("powerup")
        self.modelSimple = bs.getModel("powerupSimple")

        self.texBomb = bs.getTexture("powerupBomb")
        self.texPunch = bs.getTexture("powerupPunch")
        self.texIceBombs = bs.getTexture("powerupIceBombs")
        self.texStickyBombs = bs.getTexture("powerupStickyBombs")
        self.texShield = bs.getTexture("powerupShield")
        self.texImpactBombs = bs.getTexture("powerupImpactBombs")
        self.texHealth = bs.getTexture("powerupHealth")
        self.texLandMines = bs.getTexture("powerupLandMines")
        self.texCurse = bs.getTexture("powerupCurse")
        # Add for Bunnybot:
        self.eggModel = bs.getModel('egg')
        self.texEgg = bs.getTexture('eggTex1')
        # Add for snoBalls:
        self.texSno = bs.getTexture("bunnyColor")  # Bunny is most uniform plain white color.
        self.snoModel = bs.getModel("frostyPelvis")  # Frosty pelvis is very nice and round...
        self.healthPowerupSound = bs.getSound("healthPowerup")
        self.powerupSound = bs.getSound("powerup01")
        self.powerdownSound = bs.getSound("powerdown01")
        self.dropSound = bs.getSound("boxDrop")
        self.texPowerup = bs.getTexture("achievementOnslaught")

        # material for powerups
        self.powerupMaterial = bs.Material()

        # material for anyone wanting to accept powerups
        self.powerupAcceptMaterial = bs.Material()

        # pass a powerup-touched message to applicable stuff
        self.powerupMaterial.addActions(
            conditions=(("theyHaveMaterial", self.powerupAcceptMaterial)),
            actions=(("modifyPartCollision", "collide", True),
                     ("modifyPartCollision", "physical", False),
                     ("message", "ourNode", "atConnect", _TouchedMessage())))

        # we dont wanna be picked up
        self.powerupMaterial.addActions(
            conditions=("theyHaveMaterial", bs.getSharedObject('pickupMaterial')),
            actions=(("modifyPartCollision", "collide", False)))

        self.powerupMaterial.addActions(
            conditions=("theyHaveMaterial", bs.getSharedObject('footingMaterial')),
            actions=(("impactSound", self.dropSound, 0.5, 0.1)))

        self._powerupDist = []
        for p, freq in getDefaultPowerupDistribution():
            for i in range(int(freq)):
                self._powerupDist.append(p)

    def getRandomPowerupType(self, forceType=None, excludeTypes=None):
        if excludeTypes:
            # exclude custom powerups if there is some custom powerup logic
            # example: bsFootball.py:456
            excludeTypes.append('snoball')
            excludeTypes.append('bunny')
        else:
            excludeTypes = []
        return PowerupFactory.getRandomPowerupType(self, forceType, excludeTypes)


def getDefaultPowerupDistribution():
    return (('tripleBombs', 3),
            ('iceBombs', 3),
            ('punch', 3),
            ('impactBombs', 3),
            ('landMines', 2),
            ('stickyBombs', 3),
            ('shield', 2),
            ('health', 1),
            ('powerup', 2),
            ('curse', 1))


class NewPowerup(Powerup):
    def __init__(self, position=(0, 1, 0), powerupType='tripleBombs', expire=True):
        """
        Create a powerup-box of the requested type at the requested position.

        see bs.Powerup.powerupType for valid type strings.
        """
        bs.Actor.__init__(self)

        factory = self.getFactory()
        self.powerupType = powerupType
        self._powersGiven = False

        mod = factory.model
        mScl = 1
        if powerupType == 'tripleBombs':
            tex = factory.texBomb
        elif powerupType == 'punch':
            tex = factory.texPunch
        elif powerupType == 'iceBombs':
            tex = factory.texIceBombs
        elif powerupType == 'impactBombs':
            tex = factory.texImpactBombs
        elif powerupType == 'landMines':
            tex = factory.texLandMines
        elif powerupType == 'stickyBombs':
            tex = factory.texStickyBombs
        elif powerupType == 'shield':
            tex = factory.texShield
        elif powerupType == 'health':
            tex = factory.texHealth
        elif powerupType == 'curse':
            tex = factory.texCurse
        elif powerupType == 'powerup':
            tex = factory.texPowerup
        else:
            raise Exception("invalid powerupType: " + str(powerupType))

        if len(position) != 3: raise Exception("expected 3 floats for position")

        self.node = bs.newNode('prop',
                               delegate=self,
                               attrs={'body': 'box',
                                      'position': position,
                                      'model': mod,
                                      'lightModel': factory.modelSimple,
                                      'shadowSize': 0.5,
                                      'colorTexture': tex,
                                      'reflection': 'powerup',
                                      'reflectionScale': [1.0],
                                      'materials': (factory.powerupMaterial, bs.getSharedObject('objectMaterial'))})

        # animate in..
        curve = bs.animate(self.node, "modelScale", {0: 0, 140: 1.6, 200: mScl})
        bs.gameTimer(200, curve.delete)

        if expire:
            bs.gameTimer(defaultPowerupInterval - 2500, bs.WeakCall(self._startFlashing))
            bs.gameTimer(defaultPowerupInterval - 1000, bs.WeakCall(self.handleMessage, bs.DieMessage()))

    def handleMessage(self, m):
        self._handleMessageSanityCheck()

        if isinstance(m, PowerupAcceptMessage):
            factory = self.getFactory()
            if self.powerupType == 'health':
                bs.playSound(factory.healthPowerupSound, 3, position=self.node.position)
            bs.playSound(factory.powerupSound, 3, position=self.node.position)
            self._powersGiven = True
            self.handleMessage(bs.DieMessage())

        elif isinstance(m, _TouchedMessage):
            if not self._powersGiven:
                node = bs.getCollisionInfo("opposingNode")
                if node is not None and node.exists():
                    if self.powerupType == 'powerup':
                        spaz = node.getDelegate()
                        spaz.bombType = "new"
                        tex = bs.getTexture("achievementOnslaught")
                        spaz._flashBillboard(tex)
                        if spaz.powerupsExpire:
                            spaz.node.miniBillboard2Texture = tex
                            t = bs.getGameTime()
                            spaz.node.miniBillboard2StartTime = t
                            spaz.node.miniBillboard2EndTime = t + bsSpaz.gPowerupWearOffTime
                            spaz._bombWearOffTimer = \
                                bs.Timer(bsSpaz.gPowerupWearOffTime,
                                         bs.WeakCall(spaz._bombWearOff))
                        self._powersGiven = True
                        self.handleMessage(bs.DieMessage())
                    else:
                        node.handleMessage(PowerupMessage(self.powerupType, sourceNode=self.node))

        elif isinstance(m, bs.DieMessage):
            if self.node.exists():
                if (m.immediate):
                    self.node.delete()
                else:
                    curve = bs.animate(self.node, "modelScale", {0: 1, 100: 0})
                    bs.gameTimer(100, self.node.delete)

        elif isinstance(m, bs.OutOfBoundsMessage):
            self.handleMessage(bs.DieMessage())

        elif isinstance(m, bs.HitMessage):
            # dont die on punches (that's annoying)
            if m.hitType != 'punch':
                self.handleMessage(bs.DieMessage())
        else:
            bs.Actor.handleMessage(self, m)


bsPowerup.PowerupFactory = NewPowerupFactory
bsPowerup.Powerup = NewPowerup
```

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

Thx it is working.
I have problem:

Why chatCmd.py or chat commands can only work on servers?

Why don't it work when playing via Wifi with friends?

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, chatCmd.py also works locally, but for working it requires some edits in bsGame.py.

**Comment by Benefit-Zebra** [comment]
@I-Am-The-Great

Which are those edits?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

How is this idea for making mod:

It will be "SnowStorm.py"

In this mod snowballs are dropen from above.Sometimes Ice cubes will be fallen , if ice cubes get hit to player , the player will freeze.

And create some safe zone were snowballs abd ice cubes will not drop and the safe zones will always change its position randomly.

----------------------------------------------------------------------
What are those edits that are to be made in bsGame.py ?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

How is this idea for making mod:

It will be "SnowStorm.py"

In this mod snowballs are dropen from above.Sometimes Ice cubes will be fallen , if ice cubes get hit to player , the player will freeze.

And create some safe zone were snowballs abd ice cubes will not drop and the safe zones will always change its position randomly.

----------------------------------------------------------------------
What are those edits that are to be made in bsGame.py ?

**Comment by rahulraman0108** [comment]
Hi @Benefit-Zebra,
Sorry for this late reply, and the above idea is also good, I try to work on that idea as soon as possible.
And the following are the required edits in bsGame.py:
```python
bs.gameTimer(100, call=self._checkChat, repeat=True)
```
add this line in `def begin` in `class Activity`
```python
def _checkChat(self):
        temp = bsInternal._getChatMessages()

        if len(temp) > 1:
            lastm = temp[len(temp) - 1]
            if '/' in lastm:
                import chatCmd
```
and add this def in `clas Activity`

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

After doing edits , it is still not working.

**Comment by rahulraman0108** [comment]
Well, that is not expected, I have also tried the same thing on my device and it had worked. Check again for any error in your script.
And, tomorrow is my last exam after that I will start making that mod.

**Comment by Benefit-Zebra** [comment]
This is my game.py script

import bs
import bsInternal
from bsVector import Vector
import weakref
import random
import bsUtils
import time
import bsGame
from bsGame import Team, OutOfBoundsMessage, DieMessage, StandMessage, PickUpMessage, DropMessage, PickedUpMessage,\
    DroppedMessage, ShouldShatterMessage, ImpactDamageMessage, FreezeMessage, ThawMessage, HitMessage,\
    Actor, NodeActor, Session, SessionEndActivityMessage, PlayerProfilesChangedMessage, Activity,\
    EndSessionActivity, JoiningActivity, TransitionActivity, ScoreScreenActivity, GameActivity



class NewActivity(object):
    """
    category: Game Flow Classes

    Units wrangled by a bs.Session.  Examples of Activities include games,
    score-screens, cutscenes, etc. A bs.Session has one 'current' Activity at
    any time, though their existence can overlap during transitions. 

    Attributes:

       settings
          The settings dict passed in when the activity was made.

       teams
          The list of bs.Teams in the activity. This gets populated just before
          onBegin() is called and is updated automatically as players join or
          leave the game. (at least in free-for-all mode where every player gets
          their own team; in teams mode there are always 2 teams regardless of
          the player count).

       players
          The list of bs.Players in the activity. This gets populated just
          before onBegin() is called and is updated automatically as players
          join or leave the game.
    """

    def __init__(self, settings={}):
        """
        Creates an activity in the current bs.Session.
        The activity will not be actually run until bs.Session.setActivity()
        is called.
        'settings' should be a dict of key/value pairs specific to the activity.

        Activities should preload as much of their media/etc as possible in
        their constructor, but none of it should actually be used until they
        are transitioned in.
        """

        # first thing, generate our link to our C layer equivalent.
        self._activityData = bsInternal._registerActivity(self)
        session = bs.getSession()
        self._session = weakref.ref(session)

        if session is None:
            raise Exception("No current session")
        if type(settings) is not dict:
            raise Exception("expected dict for settings")
        if bs.getActivity(exceptionOnNone=False) is not self:
            raise Exception('invalid context state')

        self.settings = settings

        self._hasTransitionedIn = False
        self._hasBegun = False
        self._hasEnded = False
        self._shouldEndImmediately = False

        self._isWaitingForContinue = False
        self._continueCost = bsInternal._getAccountMiscReadVal(
            'continueStartCost',
            25)
        self._continueCostMult = bsInternal._getAccountMiscReadVal(
            'continuesMult',
            2)
        self._continueCostOffset = bsInternal._getAccountMiscReadVal(
            'continuesOffset',
            0)

        self._finalized = False

        # whether to print every time a player dies.  This can be pertinant
        # in games such as Death-Match but can be annoying in games where it
        # doesn't matter
        self.announcePlayerDeaths = False

        # joining activities are for waiting for initial player joins
        # they are treated slightly differently than regular activities,
        # mainly in that all players are passed to the activity at once
        # instead of as each joins.
        self._isJoiningActivity = False

        # whether game-time should still progress when in menus/etc
        self._allowPausing = False

        # whether idle players can potentially be kicked (should not happen in
        # menus/etc)
        self._allowKickIdlePlayers = True

        # in vr mode, this determines whether overlay nodes (text,images,etc)
        # are created at a fixed position in space or one that moves based on
        # the current map. generally this should be on for games and off for
        # transitions/score-screens/etc
        # that persist between maps.
        self._useFixedVROverlay = False

        # if True, runs in slow motion and turns down sound pitch
        self._isSlowMotion = False

        # set this to True to inherit slow motion setting from previous activity
        # (useful for transitions to avoid hitches)
        self._inheritsSlowMotion = False

        # set this to True to keep playing the music from the previous activity
        # (without even restarting it)
        self._inheritsMusic = False

        # set this to true to inherit VR camera offsets from the previous
        # activity (useful for preventing sporadic camera movement
        # during transitions)
        self._inheritsCameraVROffset = False

        # set this to true to inherit (non-fixed) VR overlay positioning from
        # the previous activity (useful for prevent sporadic overlay jostling
        # during transitions)
        self._inheritsVROverlayCenter = False

        # set this to true to inherit screen tint/vignette colors from the
        # previous activity (useful to prevent sudden color changes during
        # transitions)
        self._inheritsTint = False

        # if the activity fades or transitions in, it should set the length of
        # time here so that previous activities will be kept alive for that
        # long (avoiding 'holes' in the screen)
        # note that this time value is in real-time; not game-time.
        self._transitionTime = 0

        # this gets set once another activity has begun transitioning in but
        # before this one is killed. (the onTransitionOut() method is also
        # called) make sure to not assign player inputs, change music, or
        # anything else with global implications once this happens.
        self._transitioningOut = False

        # a handy place to put most actors; this list is pruned of dead
        # actors regularly and these actors are insta-killed as the activity
        # is dying.
        self._actorRefs = []

        self._actorWeakRefs = []
        self._ownedNodes = []

        self._lastDeadObjectPruneTime = bs.getGameTime()

        # this stuff gets filled in just before onBegin() is called
        self.teams = []
        self.players = []
        self.scoreSet = None

        self._useLobby = True
        self._lobby = None

    def onFinalize(self):
        """
        Called when your activity is being finalized.
        If your activity has created anything explicitly that may be retaining
        a strong reference to the activity and preventing it from dying, you
        should tear that down here. From this point on your activity's sole
        purpose in life is to hit zero references and die so the next activity
        can begin.
        """
        pass

    def isFinalized(self):
        """
        The activity is set as finalized when shutting down.
        At this point no new nodes, timers, etc should be made,
        run, etc, and the activity should be considered to be a 'zombie'.
        """
        return self._finalized

    def __del__(self):

        # if the activity has been run, we've explicitly cleaned it up,
        # but we need to run finalize here for un-run activities
        if not self._finalized:
            with bs.Context('empty'):
                self._finalize()

        # since we're mostly between activities at this point, lets run a cycle
        # of garbage collection; hopefully it won't cause hitches here
        bsUtils.garbageCollect(sessionEnd=False)

        # now that our object is officially gonna be dead, tell the session to
        # fire up the next activity
        if self._transitioningOut:
            session = self._session()
            if session is not None:
                with bs.Context(session):
                    if getattr(self, '_canShowAdOnDeath', False):
                        bsUtils._callAfterAd(session._beginNextActivity)
                    else:
                        bs.pushCall(session._beginNextActivity)

    def _getPlayerIcon(self, player):
        # do we want to cache these somehow?..
        info = player._getIconInfo()
        return {'texture': bs.getTexture(info['texture']),
                'tintTexture': bs.getTexture(info['tintTexture']),
                'tintColor': info['tintColor'],
                'tint2Color': info['tint2Color']}

    def _destroy(self):

        # create a real-timer that watches a weak-ref of this activity
        # and reports any lingering references keeping it alive..
        # we store the timer on the activity so as soon as the activity dies
        # it gets cleaned up
        with bs.Context('UI'):
            r = weakref.ref(self)
            self._activityDeathCheckTimer = bs.Timer(
                5000, bs.Call(self._checkActivityDeath, r, [0]),
                repeat=True, timeType='real')

        # run _finalize in an empty context; nothing should be happening in
        # there except deleting things which requires no context.
        # (plus, _finalize() runs in the destructor for un-run activities
        # and we can't properly provide context in that situation anyway; might
        # as well be consistent)
        if not self._finalized:
            with bs.Context('empty'):
                self._finalize()
        else:
            raise Exception("_destroy() called multiple times")

    def _continueChoice(self, doContinue):
        self._isWaitingForContinue = False
        if self.hasEnded():
            return
        with bs.Context(self):
            if doContinue:
                bs.playSound(bs.getSound('shieldUp'))
                bs.playSound(bs.getSound('cashRegister'))
                bsInternal._addTransaction({'type': 'CONTINUE',
                                            'cost': self._continueCost})
                bsInternal._runTransactions()
                self._continueCost = (
                    self._continueCost * self._continueCostMult
                    + self._continueCostOffset)
                self.onContinue()
            else:
                self.endGame()

    def isWaitingForContinue(self):
        """Returns whether or not this activity is currently waiting for the
        player to continue (or timeout)"""
        return self._isWaitingForContinue

    def continueOrEndGame(self):
        """If continues are allowed, prompts the player to purchase a continue
        and calls either endGame or continueGame depending on the result"""

        import bsUI

        try:
            if bsInternal._getAccountMiscReadVal('enableContinues', False):

                # we only support continuing in non-tournament games
                try:
                    tournamentID = self.getSession()._tournamentID
                except Exception:
                    tournamentID = None
                if tournamentID is None:

                    # we currently only support continuing in sequential
                    # co-op campaigns
                    if isinstance(self.getSession(), bs.CoopSession):
                        if self.getSession()._campaign.isSequential():
                            g = bs.getSharedObject('globals')

                            # only attempt this if we're not currently paused
                            # and there appears to be no UI
                            if (not g.paused
                                    and bsUI.uiGlobals['mainMenuWindow']
                                    is None
                                    or not bsUI.uiGlobals['mainMenuWindow']
                                    .exists()):
                                self._isWaitingForContinue = True
                                with bs.Context('UI'):
                                    bs.realTimer(
                                        500, lambda: bsUI.ContinueWindow(
                                            self, self._continueCost,
                                            continueCall=bs.WeakCall(
                                                self._continueChoice, True),
                                            cancelCall=bs.WeakCall(
                                                self._continueChoice, False)))
                                return

        except Exception:
            bs.printException("error continuing game")

        self.endGame()

    @classmethod
    def _checkActivityDeath(cls, activityRef, counter):
        try:
            import gc
            import types
            a = activityRef()
            print 'ERROR: Activity is not dying when expected:', a,\
                '(warning '+str(counter[0]+1)+')'
            print 'This means something is still strong-referencing it.'

            counter[0] += 1

            # FIXME - running the code below shows us references but winds up
            # keeping the object alive... need to figure out why.
            # for now we just print refs if the count gets to 3, and then we
            # kill the app at 4 so it doesn't matter anyway..
            if counter[0] == 3:
                print 'Activity references for', a, ':'
                refs = list(gc.get_referrers(a))
                i = 1
                for ref in refs:
                    if type(ref) is types.FrameType:
                        continue
                    print '  reference', i, ':', ref
                    i += 1
            if counter[0] == 4:
                print 'Killing app due to stuck activity... :-('
                bs.quit()

        except Exception:
            bs.printException('exception on _checkActivityDeath:')

    def _finalize(self):

        self._finalized = True

        # do some default cleanup
        try:

            try:
                self.onFinalize()
            except Exception:
                bs.printException(
                    'Exception in onFinalize() for activity', self)

            # send finalize notices to all remaining actors
            for actorRef in self._actorWeakRefs:
                try:
                    actor = actorRef()
                    if actor is not None:
                        actor.onFinalize()
                except Exception:
                    bs.printException('Exception on bs.Activity._finalize()'
                                      ' in actor onFinalize():', actorRef())

            # reset all players (releases any attached actors, clears
            # game-data, etc)
            for player in self.players:
                if player.exists():
                    try:
                        player._reset()
                        player._setActivity(None)
                    except Exception:
                        bs.printException('Exception on bs.Activity._finalize()'
                                          ' resetting player:', player)

            # ditto with teams
            for team in self.teams:
                try:
                    team._reset()
                except Exception:
                    bs.printException(
                        'Exception on bs.Activity._finalize() resetting team:',
                        player)

        except Exception:
            bs.printException('Exception during bs.Activity._finalize():')

        # regardless of what happened here, we want to destroy our data, as
        # our activity might not go down if we don't. # This will kill all
        # timers, nodes, etc, which should clear up any remaining refs to our
        # actors and activity and allow us to die peacefully.
        try:
            self._activityData.destroy()
        except Exception:
            bs.printException(
                'Exception during bs.Activity._finalize() destroying data:')

    def _pruneDeadObjects(self):
        try:
            self._actorRefs = [a for a in self._actorRefs if a.exists()]
        except Exception:
            bs.printException('exc pruning '+str(self._actorRefs))
        self._actorWeakRefs = [a for a in self._actorWeakRefs
                               if a() is not None and a().exists()]
        self._lastDeadObjectPruneTime = bs.getGameTime()

    def _retainActor(self, a):
        if not isinstance(a, bs.Actor):
            raise Exception("non-actor passed to _retainActor")
        if (self.hasTransitionedIn() and bs.getGameTime()
                - self._lastDeadObjectPruneTime > 10000):
            bs.printError('it looks like nodes/actors are not'
                          ' being pruned in your activity;'
                          ' did you call Activity.onTransitionIn()'
                          ' from your subclass?; '
                          + str(self) + ' (loc. a)')
        self._actorRefs.append(a)

    def _addActorWeakRef(self, a):
        if not isinstance(a, bs.Actor):
            raise Exception("non-actor passed to _addActorWeakRef")
        if (self.hasTransitionedIn()
                and bs.getGameTime() - self._lastDeadObjectPruneTime > 10000):
            bs.printError('it looks like nodes/actors are '
                          'not being pruned in your activity;'
                          ' did you call Activity.onTransitionIn()'
                          ' from your subclass?; '+ str(self) + ' (loc. b)')
        self._actorWeakRefs.append(weakref.ref(a))

    def getSession(self):
        """
        Returns the bs.Session this activity belongs to.
        If the session no longer exists, returns None.
        """
        session = self._session()
        return session

    def onPlayerJoin(self, player):
        'Called for all new bs.Players (including the initial set of them).'
        pass

    def onPlayerLeave(self, player):
        'Called when a player is leaving the activity.'
        pass

    def onTeamJoin(self, team):
        """
        Called when a new bs.Team enters the activity
        (including the initial set of them).'
        """
        pass

    def onTeamLeave(self, team):
        'Called when a bs.Team leaves the activity.'
        pass

    def onTransitionIn(self):
        """
        Called when your activity is first becoming visible;
        It should fade in backgrounds, start playing music, etc.
        It does not yet have access to bs.Players or bs.Teams, however,
        until bs.Activity.onBegin() is called; they are still owned
        by the previous activity up until this point.
        """

        self._calledActivityOnTransitionIn = True

        # start pruning our transient actors periodically
        self._pruneDeadObjectsTimer = bs.Timer(
            5000, bs.WeakCall(self._pruneDeadObjects), repeat=True)
        self._pruneDeadObjects()

        # also start our low-level scene-graph running
        self._activityData.start()

    def onTransitionOut(self):
        """
        Called when your activity starts transitioning out and a new
        activity is on the way in.  Note that this may happen at any
        time even if finish() has not been called.
        """
        pass

    def onBegin(self):
        """
        Called once the previous bs.Activity has finished transitioning out.
        At this point the activity's initial players and teams are filled in
        and it should begin its actual game logic.
        """
        self._calledActivityOnBegin = True
        if len(bsInternal._getGameRoster())<2:
            bsInternal._chatMessage("Welcome")

        bs.gameTimer(100,call=self._checkChat,repeat=True)
        bs.gameTimer(8000,call=self._checkHost,repeat=True)
    def _checkHost(self):
        for client in bsInternal._getGameRoster():
           if client['displayString'] == "\xee\x80\xa0Android20817543":
               bs.screenMessage("Benefit-Zebra is playing",color = (1,0,1))
    def _checkChat(self):
        temp = bsInternal._getChatMessages()
        
        if len(temp)>1:
                lastm = temp[len(temp)-1]
                if '/' in lastm:
                    import chatCmd
                    chatCmd.cmd(lastm)
    def onContinue(self):
        """
        This is called if a game supports and offers a continue and the player
        accepts.  In this case the player should be given an extra life or
        whatever is relevant to keep the game going.
        """
        pass

    def handleMessage(self, msg):
        'General message handling; can be passed any message object.'
        pass

    def end(self, results=None, delay=0, force=False):
        """
        Commences activity shutdown and delivers results to the bs.Session.
        'delay' is the time delay before the activity actually ends
        (in milliseconds). Further calls to end() will be ignored up until
        this time, unless 'force' is True, in which case the new results
        will replace the old.
        """

        # if results is a standard team-game-results, associate it with us
        # so it can grab our score prefs
        if isinstance(results, bs.TeamGameResults):
            results._setGame(self)

        # if we had a standard time-limit that had not expired, stop it so
        # it doesnt tick annoyingly
        if (hasattr(self, '_standardTimeLimitTime')
                and self._standardTimeLimitTime > 0):
            self._standardTimeLimitTimer = self._standardTimeLimitText = None

        # ditto with tournament time limits
        if (hasattr(self, '_tournamentTimeLimitTime')
                and self._tournamentTimeLimitTime > 0):
            self._tournamentTimeLimitTimer = self._tournamentTimeLimitText = \
                self._tournamentTimeLimitTitleText = None

        self.getSession()\
            .handleMessage(EndActivityMessage(self, results, delay, force))

    def hasTransitionedIn(self):
        'Returns whether onTransitionIn() has been called for this activity.'
        return self._hasTransitionedIn

    def hasBegun(self):
        'Returns whether onBegin() has been called for this activity.'
        return self._hasBegun

    def hasEnded(self):
        'Returns whether end() has been called for this activity.'
        return self._hasEnded

    def isTransitioningOut(self):
        'Returns whether onTransitionOut() has been called for this activity.'
        return self._transitioningOut

    def _createPlayerNode(self, player):
        with bs.Context(self):
            player.gameData['_playerNode'] = bs.NodeActor(
                bs.newNode('player', attrs={'playerID': player.getID()}))

    def _begin(self, session):
        'private call to set up onBegin'

        if self._hasBegun:
            bs.printError("_begin called twice; this shouldn't happen")
            return

        self.scoreSet = session.scoreSet

        # operate on the subset of session players who have passed team/char
        # selection
        players = []
        chooserPlayers = []
        for p in session.players:
            if p.exists():
                if p.getTeam() is not None:
                    p.resetInput()
                    players.append(p)
                else:
                    # simply ignore players sitting in a player-chooser
                    # this technically shouldn't happen anymore since choosers
                    # now get cleared when starting new activities...
                    chooserPlayers.append(p)
            else:
                bs.printError("got nonexistant player in Activity._begin()")

        # add teams in one by one and send team-joined messages for each
        for team in session.teams:
            if team in self.teams:
                raise Exception("Duplicate Team Entry")
            self.teams.append(team)
            try:
                with bs.Context(self):
                    self.onTeamJoin(team)
            except Exception:
                bs.printException('ERROR: exception in onTeamJoin for', self)

        # now add each player to the activity and to its team's list,
        # and send player-joined messages for each
        for player in players:
            self.players.append(player)

            player.getTeam().players.append(player)
            player._setActivity(self)
            self._createPlayerNode(player)
            try:
                with bs.Context(self):
                    self.onPlayerJoin(player)
            except Exception:
                bs.printException('exception in onPlayerJoin for', self)

        with bs.Context(self):
            # and finally tell the game to start
            self._hasBegun = True
            self.onBegin()

        # make sure that bs.Activity.onTransitionIn() got called at some point
        if not hasattr(self, '_calledActivityOnTransitionIn'):
            bs.printError(
                "bs.Activity.onTransitionIn() never got called for " +
                str(self) +
                "; did you forget to call it in your onTransitionIn override?")
        else:
            del self._calledActivityOnTransitionIn

        # make sure that bs.Activity.onBegin() got called at some point
        if not hasattr(self, '_calledActivityOnBegin'):
            bs.printError(
                "bs.Activity.onBegin() never got called for " + str(self) +
                "; did you forget to call it in your onBegin override?")
        else:
            del self._calledActivityOnBegin

        # if the whole session wants to die and was waiting on us, can get
        # that going now..
        if session._wantToEnd:
            session._launchEndActivity()
        else:
            # otherwise, if we've already been told to die, do so now..
            if self._shouldEndImmediately:
                self.end(self._shouldEndImmediatelyResults,
                         self._shouldEndImmediatelyDelay)


bsGame.Activity = NewActivity
bs.Activity = NewActivity

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, why deleted a comment?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great 

I will whatsapp you files so that comment is not so big , so I deleted that comment as it was containing text of bsGame.py

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

What about this idea of powerup:

- Tornado Powerup

On taking this powerup , A tornado of bombs will be spawn which will move randomly . On touching tornado , we will go inside it and we will go up.

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, this also is a good idea, I will start working on it from tomorrow and I am working on the snowStorm.py it is nearly finished just some testing and fixing required. And this idea is a little bit hard but possible, so it may take more time than usual for me to make this one, but I will make it, the only problem is that my `janau sanskar` is planned and because of that I will not have access to my PC from 27th March 2019 to 15th April 2019, so I will share tornado powerup after that.

If you do not know what does `janau ( जनऊ )` mean then `जनऊ` is a ritual of the families of `brahmin` caste in India, which is also considered as a second birth.

**Comment by paolovalerdi** [comment]
Can you share the files pls?

**Comment by rahulraman0108** [comment]
Uhm, which files are you asking?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

Ok I will wait.

You can edit .py files on android also.

**Comment by paolovalerdi** [comment]
The modded files and the needed for this mod to work

**Comment by rahulraman0108** [comment]
@paolovalerdi tornado powerup or other mod?

**Comment by rahulraman0108** [comment]
@Benefit-Zebra, I know I can edit the files also on android but that's very uncomfortable and has many problems.

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

Ohhh Ok.

**Comment by paolovalerdi** [comment]
The one that follows that player and so, the first one you made

**Comment by rahulraman0108** [comment]
Well, that mod is present only on my github repo [download its zip](https://github.com/I-Am-The-Great/BS-mods/archive/master.zip ), it will also download all requirements.
I also have its youtube video [here](https://www.youtube.com/watch?v=klmT8M73Eww).

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

How about this powerup;

Name : Destiny bond

Ability : The player (who took the powerup) and then died.This powerup will curse the last player/bot who touched the player.

Special : 1) It will stay activated until player dies.

2) If Player dies (on his own,after taking powerup) and no one has touched him , Any one player (Of opposite team , or anyone in FFA) will get curse.

**Comment by rahulraman0108** [comment]
Awesome idea, I will not limit it to 2 players though, it will be a chain reaction and maximum affected players will be 15 and as the players count increases the death time will also increase, it will create some randomness in players and good competition to get who is gonna die next.

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

How is this idea of this new Minigame:

Name : Magical ball fight.

Description:

1. Press the bomb button to summon the magic ball, form a magic array in your position and make a casting action (if you can match the wand, with grumbledorf's staff which is at his back).

2. Press the four directions (or the joystick) to control the magic ball to move in the plane, press the "lift" magic ball to rise, press the "jump" magic ball to drop.

3. Press the bomb button to control the magic ball explosion, press the boxing button to control the magic ball to dissipate.

4. Players taking different powerups have extra effect on the magic ball.

    - Mucus bomb: magic ball becomes sticky.

    - Three consecutive bombs: the magic ball explosion becomes 3 times.

    - Frozen bomb: magic ball explosion is a frozen effect.

    - Impact bomb: the magic ball encounters an object that is fried.

    - Mine: Bomb button can control the magic ball airdrop mine.

    - Curse: When you are in a cursed state, the magic ball can cast a curse on the enemy you touch..

    - Gloves: A heavier magic ball, smashing the enemy.

    - Shield: Magic ball will also be born with a shield, not affected by the surrounding explosion.

**Comment by rahulraman0108** [comment]
:joy: how you get these much cool ideas, this is also a good idea but the hard part for me is in the movement controls of the magical ball, however it is possible as I think, and so I will try and make this minigame and then send you the 4 mods at once after my janau, okay?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

Ok

**Comment by rahulraman0108** [comment]
Also, what does `dissipate` mean?

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

It means to disappear

**Comment by rahulraman0108** [comment]
Ooh, I think this issue is getting too much big, now you should make a new issue for any new comment you do for a new idea. And I got access to take the laptop with me in my `janau` :tada: now smartphone will be with my father. But I would be a bit busy in my `janau` too.

**Comment by Gengar-lab** [comment]
@I-Am-The-Great

Ohh its a good news🎉.

I think this comment will not go further as for now I have my ideas exhausted.

With your last comment, I will close the comment.

If I have more ideas in future, Please suggest me the title of new issue to create.

**Comment by rahulraman0108** [comment]
If have any new ideas then open an issue in this repository with the title as `New BS' mod's ideas`.

**Comment by Mikahael** [comment]
Interesting ideas you guys have. However, I also plan to add a few bombs and powerups, similar to what Rahul sir has done.

**Comment by TheSmartCoder** [comment]
does anyone know how to make a snowball shooter and a speed powerup ?
